### PR TITLE
use v3io/frames

### PR DESCRIPTION
**Pull request by talIguaz**

### CODE REVIEW

File: pkg/pquerier/frames.go [hunk 1]
```diff
@@ -332,6 +347,54 @@ func (d *dataFrame) TimeSeries(i int) (utils.Series, error) {
...
+// First do all the concrete columns and then the virtual who are dependant on the concrete.
+func (d *dataFrame) finishAllColumns() error {
+
+	// Marking as Deleted all the indexes that has no data.
+	for i, hasData := range d.nonEmptyRowsIndicator {
+		if !hasData {
+			for _, col := range d.columns {
+				_ = col.Delete(i)
+			}
+			d.index.Delete(i)
```
Comment (gtopper) [inline pkg/pquerier/frames.go]: Unhandled error.

File: pkg/pquerier/frames.go [hunk 2]
```diff
@@ -31,8 +22,16 @@ type frameIterator struct {
 }
 
 // create new frame set iterator, frame iter has a SeriesSet interface (for Prometheus) plus columnar interfaces
-func NewFrameIterator(ctx *selectQueryContext) *frameIterator {
-	return &frameIterator{ctx: ctx, columnNum: ctx.totalColumns, setIndex: 0, seriesIndex: -1}
+func NewFrameIterator(ctx *selectQueryContext) (*frameIterator, error) {
+	if !ctx.isRawQuery() {
+		for _, f := range ctx.frameList {
+			if err := f.finishAllColumns(); err != nil {
+				return nil, err
```
Comment (gtopper) [inline pkg/pquerier/frames.go]: Shouldn't it also says which dataframe failed? (i.e. wrap the error with this information)

File: pkg/pquerier/frames.go [hunk 3]
```diff
@@ -214,10 +214,11 @@ type dataFrame struct {
...
 
-	columnsTemplates []columnMeta
-	columns          []Column
-	index            Column
-	columnByName     map[string]int // name -> index in columns
+	columnsTemplates      []columnMeta
+	columns               []Column
+	index                 Column
+	columnByName          map[string]int // name -> index in columns
+	nonEmptyRowsIndicator []bool
```
Comment (gtopper) [inline pkg/pquerier/frames.go]: Looks like indicators to me.

File: pkg/pquerier/frames.go [hunk 4]
```diff
@@ -332,6 +347,54 @@ func (d *dataFrame) TimeSeries(i int) (utils.Series, error) {
...
+		case *virtualColumn:
+			newCol.size = columnSize
+			err = col.finish()
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	d.index.finish()
```
Comment (gtopper) [inline pkg/pquerier/frames.go]: Unhandled error.

File: pkg/pquerier/frames.go [hunk 5]
```diff
@@ -332,6 +347,54 @@ func (d *dataFrame) TimeSeries(i int) (utils.Series, error) {
...
+		case *ConcreteColumn:
+			err = col.finish()
+			if columnSize == 0 {
+				columnSize = col.FramesColumn().Len()
+			} else if columnSize != col.FramesColumn().Len() {
+				return fmt.Errorf("columns length mismatch %v!=%v col=%v", columnSize, col.FramesColumn().Len(), col.Name())
+			}
+		}
+		if err != nil {
+			return err
```
Comment (gtopper) [inline pkg/pquerier/frames.go]: Maybe let the user know which column failed?

File: pkg/pquerier/frames.go [hunk 6]
```diff
@@ -332,6 +347,54 @@ func (d *dataFrame) TimeSeries(i int) (utils.Series, error) {
...
+		}
+	}
+	for _, col := range d.columns {
+		switch newCol := col.(type) {
+		case *virtualColumn:
+			newCol.size = columnSize
+			err = col.finish()
+		}
+		if err != nil {
+			return err
```
Comment (gtopper) [inline pkg/pquerier/frames.go]: Maybe let the user know which column failed?

File: pkg/pquerier/frames.go [hunk 7]
```diff
@@ -332,6 +347,54 @@ func (d *dataFrame) TimeSeries(i int) (utils.Series, error) {
...
+	for _, col := range d.columns {
+		switch col.(type) {
+		case *dataColumn:
+			err = col.finish()
+		case *ConcreteColumn:
+			err = col.finish()
+			if columnSize == 0 {
+				columnSize = col.FramesColumn().Len()
+			} else if columnSize != col.FramesColumn().Len() {
+				return fmt.Errorf("columns length mismatch %v!=%v col=%v", columnSize, col.FramesColumn().Len(), col.Name())
```
Comment (gtopper) [inline pkg/pquerier/frames.go]: Looks like "column length"

Comment (talIguaz) [reply pkg/pquerier/frames.go]: it's `columns`, since the length of 2 or more columns was not the same.

File: pkg/pquerier/frames.go [hunk 8]
```diff
@@ -665,31 +690,33 @@ type virtualColumn struct {
...
-func (c *virtualColumn) FloatAt(i int) (float64, error) {
-	if !c.isValidIndex(i) {
-		return 0, fmt.Errorf("index %d out of bounds [0:%d]", i, c.size)
+func (c *virtualColumn) finish() error {
+	data := make([]float64, c.Len())
+	var err error
+	for i := 0; i < c.Len(); i++ {
+		value, err := c.function(c.dependantColumns, i)
+		if err != nil {
+			data[i] = math.NaN()
```
Comment (gtopper) [inline pkg/pquerier/frames.go]: We're just going to throw the error away here?

File: pkg/pquerier/select.go [hunk 9]
```diff
@@ -104,8 +103,8 @@ func (queryCtx *selectQueryContext) start(parts []*partmgr.DBPartition, params *
 		queryCtx.totalColumns = queryCtx.frameList[0].Len()
 	}
 
-	frameIter := NewFrameIterator(queryCtx)
-	return frameIter, nil
+	frameIter, err := NewFrameIterator(queryCtx)
+	return frameIter, err
```
Comment (gtopper) [inline pkg/pquerier/select.go]: I think you can actually just `return NewFrameIterator(queryCtx)` in this case 😄

File: pkg/pquerier/frames.go [hunk 10]
```diff
@@ -31,8 +22,16 @@ type frameIterator struct {
 }
 
 // create new frame set iterator, frame iter has a SeriesSet interface (for Prometheus) plus columnar interfaces
-func NewFrameIterator(ctx *selectQueryContext) *frameIterator {
-	return &frameIterator{ctx: ctx, columnNum: ctx.totalColumns, setIndex: 0, seriesIndex: -1}
+func NewFrameIterator(ctx *selectQueryContext) (*frameIterator, error) {
+	if !ctx.isRawQuery() {
+		for _, f := range ctx.frameList {
+			if err := f.finishAllColumns(); err != nil {
+				return nil, errors.Wrap(err, fmt.Sprintf("failed to create columns for DF=%v", f.Labels()))
```
Comment (gtopper) [inline pkg/pquerier/frames.go]: `Wrapf` 😄

File: pkg/pquerier/frames.go [hunk 11]
```diff
@@ -332,6 +347,54 @@ func (d *dataFrame) TimeSeries(i int) (utils.Series, error) {
 	}
 }
 
+// Creates Frames.columns out of tsdb columns.
+// First do all the concrete columns and then the virtual who are dependant on the concrete.
+func (d *dataFrame) finishAllColumns() error {
+
+	// Marking as Deleted all the indexes that has no data.
```
Comment (gtopper) [inline pkg/pquerier/frames.go]: 🇬🇧❓

File: pkg/pquerier/query_integration_test.go [hunk 12]
```diff
@@ -1580,15 +1583,17 @@ func (suite *testQuerySuite) TestVariantTypeQueryWithDataFrame() {
...
-		in := frame.Index()
-		cols := frame.Columns()
+		frame, err := iter.GetFrame()
+		suite.NoError(err)
+		indexCol := frame.Indices()[0] // in tsdb we have only one index
 
-		for i := 0; i < frame.Index().Len(); i++ {
-			t, _ := in.TimeAt(i)
+		for i := 0; i < indexCol.Len(); i++ {
+			t, _ := indexCol.TimeAt(i)
```
Comment (gtopper) [inline pkg/pquerier/query_integration_test.go]: Check `NoError`?

File: pkg/pquerier/query_integration_test.go [hunk 13]
```diff
@@ -2255,23 +2270,25 @@ func (suite *testQuerySuite) TestStringAndFloatMetricsDataframe() {
...
+				column, err := frame.Column(columnName)
+				suite.NoError(err)
+				if column.DType() == frames.FloatType {
 					v, _ = column.FloatAt(i)
-				} else if column.DType() == pquerier.StringType {
+				} else if column.DType() == frames.StringType {
 					v, _ = column.StringAt(i)
 				} else {
-					suite.Failf("column type is not as expected: %v", column.DType().String())
+					suite.Fail(fmt.Sprintf("column type is not as expected: %v", column.DType()))
```
Comment (gtopper) [inline pkg/pquerier/query_integration_test.go]: What was wrong with `Failf`? :)

**Pull request merged**